### PR TITLE
Pull Request: Correlation Remover

### PR DESCRIPTION
**Pull request by koaning**
@MiroDudik it's taken me a while but I've started the PR just to get the ball rolling again. 

This PR hopes to adress https://github.com/fairlearn/fairlearn/issues/430.

### CODE REVIEW

**Comment by koaning** [comment]
I now realise that we also had https://github.com/fairlearn/fairlearn/pull/464. I'll copy relevant parts into this PR and I'll close the other one later.

**Review by romanlutz — COMMENTED** [review]
I'm super excited that we're (finally) starting a preprocessing module :-) Thanks @koaning !

How do you feel about adding a user guide section, perhaps with some of the charts you had in your notebook?

File: fairlearn/preprocessing/_correlation_remover.py
Comment (koaning) [reply fairlearn/preprocessing/_correlation_remover.py:58]: What needs to happen to the output? If it is data-frame in does it have to be data-frame out? I might have a preference to cast it into a numpy array internally to ensure that the maintenance stays easy.

File: fairlearn/preprocessing/_correlation_remover.py
Comment (koaning) [reply fairlearn/preprocessing/_correlation_remover.py:64]: The `y=None` is indeed for consistency. Just to confirm though, the `.fit()` method inside fairlearn pipelines send `sensitive_features` as well?

Comment (romanlutz) [reply fairlearn/preprocessing/_correlation_remover.py:64]: `fit` in Fairlearn has `sensitive_features`, but I'm not 100% what you mean by Fairlearn pipelines.

File: fairlearn/preprocessing/_correlation_remover.py
Comment (romanlutz) [reply fairlearn/preprocessing/_correlation_remover.py:58]: I'm perfectly happy with it being numpy array as the return type, but it should handle both input data types. In other places we've used input validation from scikit-learn to check data types. The documentation actually doesn't tell you in our `predict` methods, just says "vector" which I find problematic... https://fairlearn.github.io/api_reference/fairlearn.reductions.html#fairlearn.reductions.ExponentiatedGradient

File: fairlearn/preprocessing/_correlation_remover.py
Comment (koaning) [reply fairlearn/preprocessing/_correlation_remover.py:64]: Ah right. This might be a larger issue then. 

If we want the tools here to be scikit-learn compatible then this preprocessor needs to be able to the be used in a `Pipeline`. For example, a standard scaler might be applied via; 

```python
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression

pipe = Pipeline([
  ('prep', StandardScaler()), 
  ('model', LogisticRegression())
]) 

# The scikit-learn pipeline here does not have `sensitive_features`. 
pipe.fit(X, y) 
```

If we want `CorrelationRemover` to be scikit-learn compatible with regards to the `Pipeline` then we cannot assume to have `sensitive_features` available in the `.fit()`.

Comment (romanlutz) [reply fairlearn/preprocessing/_correlation_remover.py:64]: This goes back to a larger discussion on this topic that @adrinjalali started some time back. If I remember correctly there will be some way to do this with scikit-learn (similar to how you can pass sample_weight), but I'm afraid I'm lacking the context @adrinjalali has.

Comment (adrinjalali) [reply fairlearn/preprocessing/_correlation_remover.py:64]: You can pass arbitrary parameters to `fit` in a `Pipeline` via `fit_params`: https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html#sklearn.pipeline.Pipeline.fit

So in this case, it' be like `pipeline.fit(X, y, correlationremover__sensitive_features=blahblah)`

Comment (MiroDudik) [reply fairlearn/preprocessing/_correlation_remover.py:64]: In the context of this PR, we should just go with the original proposal, which is:
* provide something like `sensitive_feature_ids` / `sensitive_feature_columns` (pandas terminology?) to the constructor and then we can work with `fit(X,y)` and `predict(X)`

A seeming alternative would be:
* do it similarly to our post-processing step where we have `fit(X,y,sensitive_features=sf)` and also `predict(X,sensitive_features=sf)`.

But this is not compatible with `sklearn.Pipeline`, because I think it only allows `predict_params` at the [very last stage of the pipeline](https://github.com/scikit-learn/scikit-learn/blob/0fb307bf3/sklearn/pipeline.py#L381) (whereas we need them at the beginning).

File: fairlearn/preprocessing/_correlation_remover.py
Comment (MiroDudik) [inline fairlearn/preprocessing/_correlation_remover.py:53]: I don't think that we need or use `center` anywhere any more, let's drop.

File: fairlearn/preprocessing/_correlation_remover.py
Comment (MiroDudik) [inline fairlearn/preprocessing/_correlation_remover.py:10]: I personally like the name `CorrelationRemover`, but I wanted to bring back one alternative, which I think is a bit worse, but worth mentioning: `LinearDependenceRemover`.

Both names are accurate descriptions of the transformer. `CorrelationRemover` is more compact, but one possible disadvantage is that many people might think that it removes the **statistical dependence** (rather than just correlation). Calling it `LinearDependenceRemover` would deal with that... but the name is a bit awkward. Thoughts?

File: fairlearn/preprocessing/_correlation_remover.py
Comment (MiroDudik) [reply fairlearn/preprocessing/_correlation_remover.py:58]: Agreed re. accepting both `pandas.DataFrame` and `numpy.ndarray` as inputs, but only producing `numpy.ndarray` as output. This is consistent with the [sklearn API](https://scikit-learn.org/stable/glossary.html#term-transform).

**Comment by MiroDudik** [comment]
@koaning: let us know if you are blocked anywhere. As far as I'm concerned, this is good to go with the current comments addressed.

File: fairlearn/preprocessing/_correlation_remover.py
Comment (MiroDudik) [inline fairlearn/preprocessing/_correlation_remover.py:30]: I think we should say something like:
```
Note that the lack of correlation does not imply anything about statistical dependence. Therefore, we expect this to be most appropriate as a preprocessing step for (generalized) linear models.
```

File: fairlearn/preprocessing/_correlation_remover.py
Comment (adrinjalali) [reply fairlearn/preprocessing/_correlation_remover.py:64]: I see, it's true that we're not accepting extra params during `predict` at the moment, and it needs to be fixed. We can't pass those in the constructor because the data passed during `predict` can be very different than the one passed during `fit`. And in reality, we will not have access to the `predict` data during `fit`, therefore we can't pass that to the constructor (think a model deployed  to do predictions in real-time).

File: fairlearn/preprocessing/_correlation_remover.py
Comment (koaning) [reply fairlearn/preprocessing/_correlation_remover.py:10]: I'm cool with either.

File: fairlearn/preprocessing/_correlation_remover.py
Comment (koaning) [reply fairlearn/preprocessing/_correlation_remover.py:64]: So I can keep it at `sensitive_feature_ids` as constructor input? Presumably in such a way that it is compatible for both DataFrame and numpy?

File: fairlearn/preprocessing/_correlation_remover.py
Comment (romanlutz) [reply fairlearn/preprocessing/_correlation_remover.py:10]: I don't think `LinearDependenceRemover` is awkward, and it's a bit more explicit in saying what it does and that it's not a solution for all problems one could ever have. That said, this is not a hill I'm willing to die on :-)

File: fairlearn/preprocessing/_correlation_remover.py
Comment (MiroDudik) [inline fairlearn/preprocessing/_correlation_remover.py:61]: This does not work when `sensitive` is a list of strings and `X` is a `pandas.DataFrame`.

File: fairlearn/preprocessing/_correlation_remover.py
Comment (MiroDudik) [reply fairlearn/preprocessing/_correlation_remover.py:64]: Yes! Please keep `sensitive_feature_ids` as the constructor argument--with either a list of strings or a list of ints.

File: fairlearn/preprocessing/_correlation_remover.py
Comment (MiroDudik) [reply fairlearn/preprocessing/_correlation_remover.py:53]: Actually, on the second thought I'm in favor of the second option:
```python
def __init__(self, *, sensitive_feature_ids, alpha=1):
```
because it makes our API more future-proof. For example, if we want to add `control_feature_ids` as we did in some of our other metrics.

**Comment by koaning** [comment]
Anybody else getting segfaults when trying to build the docs locally? 

```
make: *** [doc] Segmentation fault: 11. [100%] plot_grid_search_census.py 
```

It also seems to need a manual `pip install rai_core_flask`. Should it perhaps be added to `dev-requirements.txt`?

**Comment by koaning** [comment]
I'm getting near but the flake8 in the CI is showing some messages that I don't have locally. 

```
/bin/bash --noprofile --norc /home/vsts/work/_temp/7b79ef4f-5386-47b7-b7ca-9bcdbff2a0bf.sh
./test/unit/preprocessing/__init__.py:1:2: C801 Copyright notice not present.
./test/unit/preprocessing/linear_dep_remover/__init__.py:1:2: C801 Copyright notice not present.
./test/unit/preprocessing/linear_dep_remover/test_sklearn_compat.py:1:2: C801 Copyright notice not present.
./fairlearn/preprocessing/__init__.py:1:2: C801 Copyright notice not present.
./fairlearn/preprocessing/_linear_dep_remover.py:1:2: C801 Copyright notice not present.
```

**Comment by romanlutz** [comment]
> I'm getting near but the flake8 in the CI is showing some messages that I don't have locally.
> 
> ```
> /bin/bash --noprofile --norc /home/vsts/work/_temp/7b79ef4f-5386-47b7-b7ca-9bcdbff2a0bf.sh
> ./test/unit/preprocessing/__init__.py:1:2: C801 Copyright notice not present.
> ./test/unit/preprocessing/linear_dep_remover/__init__.py:1:2: C801 Copyright notice not present.
> ./test/unit/preprocessing/linear_dep_remover/test_sklearn_compat.py:1:2: C801 Copyright notice not present.
> ./fairlearn/preprocessing/__init__.py:1:2: C801 Copyright notice not present.
> ./fairlearn/preprocessing/_linear_dep_remover.py:1:2: C801 Copyright notice not present.
> ```
> 
> @romanlutz any advice?

@koaning I'm afraid flake8 is right. You can fix all of these by simply copying the copyright notice from another file and pasting it at the beginning. It should complain about that locally as well, provided you have your branch synced with the latest version. GitHub does that automagically I believe.

**Comment by romanlutz** [comment]
> Anybody else getting segfaults when trying to build the docs locally?
> 
> ```
> make: *** [doc] Segmentation fault: 11. [100%] plot_grid_search_census.py 
> ```
> 
> It also seems to need a manual `pip install rai_core_flask`. Should it perhaps be added to `dev-requirements.txt`?

That was rolled back a few weeks ago. You should probably sync with master. Otherwise I completely agree that dev requirements need to be captured in `dev-requirements.txt`.

I'm not getting segfaults. @riedgar-ms have you seen this with doc builds?

**Comment by koaning** [comment]
@romanlutz I think what went wrong is that I had the tools installed but not the correct version. Since the `dev-requirements.txt` file does not list a minimum version for the tools my system didn't get the nudge to update the minimum required version. I just installed with `python -m pip install --upgrade dev-requirements.txt` and that did cause the flake8 to start complaining.

**Comment by koaning** [comment]
That said, it's starting to look green. The only thing is that I cannot seem to build the docs locally due to the segfault.

```
make: *** [doc] Segmentation fault: 11. [100%] plot_grid_search_census.py                                                                                                                                                              
```

**Comment by romanlutz** [comment]
> @romanlutz I think what went wrong is that I had the tools installed but not the correct version. Since the `dev-requirements.txt` file does not list a minimum version for the tools my system didn't get the nudge to update the minimum required version. I just installed with `python -m pip install --upgrade dev-requirements.txt` and that did cause the flake8 to start complaining. 

Interesting! Let's set minimum versions for packages. @riedgar-ms @hildeweerts @adrinjalali @MiroDudik wdyt?

**Comment by riedgar-ms** [comment]
Does setting minimum versions in the requirements files actually help? I have a vague memory of being caught once where a requirements file specified a minimum, but pip saw the package was present, and didn't check the version.

If my memory is in error, I have no objection to adding minimum versions too.

**Review by romanlutz — APPROVED** [review]
There's a tiny issue with the docs, but other than that this looks good to me, so I'm approving this. @riedgar-ms and/or @MiroDudik please take a look as well and approve/complete when the issue is fixed assuming you have no other requests.

See the rendered version here: https://2071-133444044-gh.circle-artifacts.com/0/docs/_build/html/api_reference/fairlearn.preprocessing.html

image

I believe this has to do with the format needing to be
```
Parameters
----------
param_name : type
    description
```
and that indent in the second line was missing.

**Comment by adrinjalali** [comment]
re: minimum requirements, they should be in `setup.py` for them to actually take effect when the user installs either from source or a binary. We could also put them in `pyproject.toml` for pip to better understand it, and read those requirements from that file in `setup.py` to have them synced. We can even condition the minimum requirements on architecture and python version there.

This doesn't include the optional requirements though. For those we could have them in `requirements.txt` or any other place in the documentation.

**Comment by koaning** [comment]
Docstring should be fixed now. @romanlutz

**Comment by romanlutz** [comment]
> Docstring should be fixed now. @romanlutz

Thank you! @MiroDudik feel free to complete if you're fine with the changes.

I did notice that there's no update to CHANGES.md, but if @koaning doesn't get to it we can always add it in a separate PR or push it on top of this one before completing.

File: fairlearn/preprocessing/_linear_dep_remover.py
Comment (MiroDudik) [inline fairlearn/preprocessing/_linear_dep_remover.py:73]: this is a debugging left-over i think.

File: fairlearn/preprocessing/_linear_dep_remover.py
Comment (MiroDudik) [inline fairlearn/preprocessing/_linear_dep_remover.py:92]: add `"sensitive_mean_"`

File: fairlearn/preprocessing/_linear_dep_remover.py
Comment (MiroDudik) [inline fairlearn/preprocessing/_linear_dep_remover.py:95]:
```suggestion
                f"The trained data has {self.X_shape_[1]} features while this dataset has {X.shape[1]}."
```

File: fairlearn/preprocessing/_linear_dep_remover.py
Comment (MiroDudik) [inline fairlearn/preprocessing/_linear_dep_remover.py:79]:
```suggestion
        """Learn the projection required to make the dataset uncorrelated with sensitive columns."""
```

File: fairlearn/preprocessing/_linear_dep_remover.py
Comment (MiroDudik) [inline fairlearn/preprocessing/_linear_dep_remover.py:90]:
```suggestion
        """Transform X by applying the correlation remover."""
```

**Review by MiroDudik — APPROVED** [review]
This is good to merge when:
* the remaining suggestions are addressed
* we go back to `CorrelationRemover`

@romanlutz agreed to deal with this on merge, so @koaning -- you don't need to take action on these.

By the way, I apologize for going back and forth on the naming from `LinearDependenceRemover` to `CorrelationRemover`, but I just realized that because of the centering step, it is not correct to call this `LinearDependenceRemover`, so let's go back to `CorrelationRemover`.

**Comment by koaning** [comment]
OK. I think the DCO is now broken because I've accepted @MiroDudik 's comments via github. The command-line tool doesn't seem to be able to "fix" it. @romanlutz any help?

**Comment by romanlutz** [comment]
> OK. I think the DCO is now broken because I've accepted @MiroDudik 's comments via github. The command-line tool doesn't seem to be able to "fix" it. @romanlutz any help?

Yes it does that, because the suggestions don't have DCO sign-off. DCO is really a high friction topic and perhaps we should find a better solution (suggestions highly welcome! Maybe I should start an issue...). 
Let me quickly go through Miro's points and try to fix it.

**Comment by adrinjalali** [comment]
I thought changing the governance would also mean removing the DCO requirement.

**Comment by romanlutz** [comment]
> I thought changing the governance would also mean removing the DCO requirement.

We certainly need some sort of agreement for contributions, don't we? I'm not well versed in open source from a legal perspective, so feel free to explain better ways of doing this. You know what, let me start an issue... Let's take this there.

**Comment by romanlutz** [comment]
> > OK. I think the DCO is now broken because I've accepted @MiroDudik 's comments via github. The command-line tool doesn't seem to be able to "fix" it. @romanlutz any help?
> 
> Yes it does that, because the suggestions don't have DCO sign-off. DCO is really a high friction topic and perhaps we should find a better solution (suggestions highly welcome! Maybe I should start an issue...).
> Let me quickly go through Miro's points and try to fix it.

As it happens so often the steps outlined in the DCO "pipeline" don't even work. I tried it several times without success. Since @koaning  signed off on everything but the suggestion from Miro I consider that sufficient and have set the DCO to pass. If there are additional commits that has to be manually repeated. Again, I hope we can find a better solution via #650 for this problem. Apologies again for the waste of time.

**Comment by koaning** [comment]
It's nice to know that my little technique has a place to live. 

The work on this PR started in [May](https://github.com/fairlearn/fairlearn/issues/430) and it's December now. Looking back, I figured it might be good to share some notes/observations on my end. There's a lot in this project that I think should be addressed;

- There's still a big Microsoft-branded feel to the docs/project, which I feel the project would be better off without.
- The docs deserve better explanations of the methods, linking to an academic article as an explanation isn't very inclusive.
- Getting something merged feels harder than it should be (segfaults on building docs, DCO, etc). 

But all of this said ...  I do have a nice observation now that this PR is done. In the end, this "trick" that I came up with two years ago got a proper review here, and in doing so ... it got better. Like; properly better! With @MiroDudik's feedback, I like to think the preprocessor is not only much faster but also numerically more stable than it was before. It's exactly this phenomenon that makes for good results with open-source tools.

I might return later in time to contribute more, but for now, I need to focus more on some other projects that might help out in this space. 

- [human learn](https://koaning.github.io/human-learn/) allows you to more flexibly declare models using business rules, forcing the user to explore the data and what is in there. It also comes with a nice-drawing feature that can be used for [bulk labelling](https://www.youtube.com/watch?v=YsMoGd7sYMQ&list=PL75e0qA87dlG-za8eLI6t0_Pbxafk-cxb&index=31&ab_channel=Rasa). We simply need to have more data scientists willing to be confronted with their own data.  
- [whatlies](https://github.com/RasaHQ/whatlies) allows you to visualize properties in word embedding, which should help you explore the bias that is in them 
- [deon](https://deon.drivendata.org/) is an awesome checklist with inspirational stories of unintended side effects to watch out for. I've been meaning to contribute some more examples there. 
- [koaning.io](https://koaning.io/posts/mean-squared-terror/) will keep collecting failure scenarios for the standard grid-search approach to ML. 
- [algorithm whiteboard](https://www.youtube.com/watch?v=Czto6GzJah8&list=PL75e0qA87dlG-za8eLI6t0_Pbxafk-cxb&index=32&ab_channel=Rasa) will remain a youtube channel that explains algorithms in NLP, as well as their unwanted side-effects. In particular, here's my explainer of why dense debiasing-techniques for word-embeddings [don't work](https://www.youtube.com/watch?v=2ROP1QFKsqc&list=PL75e0qA87dlG-za8eLI6t0_Pbxafk-cxb&index=26&ab_channel=Rasa). 
- [tokenwiser](https://github.com/koaning/tokenwiser) will be my attempt at pretraining embeddings by only considering linguistic properties. Who knows, maybe someday I'll get to brag about a discovery.

**Comment by hildeweerts** [comment]
Thank you so much your feedback @koaning, this is very helpful!

**Comment by adrinjalali** [comment]
> There's still a big Microsoft-branded feel to the docs/project, which I feel the project would be better off without.

I agree. With the project becoming an independent entity, this will be much easier to achieve.

> The dashboard (last time I tried running fairlearn locally) requires a whole lot of jupyter plugins and it really feels like a barrier of entry.

The dashboard is to be removed from the next release, so this should be solved.

> The docs deserve better explanations of the methods, linking to an academic article as an explanation isn't very inclusive.

+100

> Getting something merged feels harder than it should be (segfaults on building docs, DCO, etc).

For the segfaults and build issues it would be nice if you could open separate issues when they happen. They need to be fixed. As the DCO issue, I totally agree it should be removed.

Thanks for all your contributions @koaning :)

**Comment by romanlutz** [comment]
Thanks a lot for sharing @koaning ! I agree with everything written by @hildeweerts and @adrinjalali . In a way, I feel quite encouraged because these are all things we can actually improve, and for most of them we're already in the process of changing it (DCO, governance).

Documentation is something we all should think about addressing. I'm wondering whether it makes sense for me to adopt a 30-minute-a-day documentation habit for a while just to make sure we actually get to it.

I'm glad your review experience with @MiroDudik was so positive :-) Thanks for your patience throughout this process, and for sharing this honest feedback.

**Comment by koaning** [comment]
I might be able to contribute a https://calmcode.io series that can be hosted on the docs at some point. I would prefer not to have to read an article of maths before being able to work on it though. 😅

**Comment by romanlutz** [comment]
@koaning this sounds great. If you let me know which ones specifically you need clarifications on first we can prioritize those.

**Comment by MiroDudik** [comment]
@koaning : So glad it's in and apologies for the various delays along the way! I'm very happy with the result. And as others said, thanks so much for the feedback!

**Pull request merged**
